Add Crown Jewel resource

jakule opened the pull request:
Add Crown Jewel resource described here https://github.com/gravitational/access-graph/blob/main/rfd/0005-crown-jewel.md

Depends on https://github.com/gravitational/teleport/pull/41168

changelog: Add support for crown jewel resource type

tigrato committed:
Update labels datatype in Teleport crownjewel proto

This commit modifies the labels from a map to repeated TeleportLabel in CrownJewel proto, and accordingly updates the getter function in proto generated code. The newly introduced TeleportLabel type contains a Key and repeated values for improved flexibility.

Clean up

Update and restructure CrownJewel service proto

The code modifications include renaming and restructuring several types and altering various functions in the CrownJewel service proto. Major changes involve renaming DeleteCrownJewel

Rename 'kind' to 'kinds' in crownjewel API

This commit changes the 'kind' field to 'kinds' in the crownjewel API to better reflect its purpose of holding multiple resource types. Changes are made across several files - 'crownjewel.go', 'crownjewel.pb.go' and 'crownjewel.proto'.

After rebase fixes

Remove CrownJewel type from several files

This commit represents the removal of the CrownJewel type from several files. CrownJewel type references are deleted in certain Go files, and imports of CrownJewel in multiple files are being adjusted. The specifics of why this type is removed is not evident from the diffs, but the deletions will affect the related functionalities.

Add CrownJewel update functionality

This update enables the 'UpdateCrownJewel' functionality. If a 'CrownJewel' resource already exists, it can now be updated instead of returning an error. This enhances flexibility when dealing with resource management. It also introduces the ARN field into the 'AWSMatcher' service.

Update CrownJewel service with pagination and improve matchers

This update adds pagination to the CrownJewel service for more efficient data handling. It also improves the matching capabilities with the inclusion of Teleport and AWS matchers and allows for selective querying with these matchers. Additionally, it includes various changes to structure and naming for better clarity and ease of use.

Refactor CrownJewel protos to use ResourceHeader

This refactoring replaces the previous structure of CrownJewel proto's fields Kind, Sub_kind, Version, and Metadata with a single object called Header (ResourceHeader). It simplifies the code structure and harmonizes it with other protos. Now, Header contains the essential attributes that were previously distributed across multiple separate fields.

Add GetCrownJewels method to cache and accessgraph

The function GetCrownJewels has been added to the cache and the accessgraph. It retrieves all CrownJewel instances from the cache. Additionally, CrownJewel retrieval functionality has been removed from auth server and existing usages have been refactored.

Implement Crown Jewel creation feature

Updated the method `createCrownJewel` in the resources command to handle Crown Jewel creation. Introduced a new gRPC service `crownjewelv1` to handle client-server interactions for Crown Jewel objects. Crown Jewel expiry configuration has now also been enabled.

Add missing files

I have no words...

Update authentication and event handling

Refactored the argument from 'auth.ClientI' to '*auth.Client' in 'createCrownJewel' function in 'resource_command.go'. Removed unused import 'durationpb' in 'auth_with_roles.go'. Added a new event code 'AccessGraphAccessPathChangedCode' in 'codes.go' and 'AccessGraphAccessPathChanged' in 'api.go' for better monitoring and logging of access path changes.

Add access path changed event and relevant changes.

 update api

add grpc

update api

add cache support

add protos

jakule committed:
Fixes after rebase

jakule committed:
Refactor Crown Jewel resource methods

Updated API methods related to the Crown Jewel resource to improve clarity and consistency. Renamed `GetCrownJewels` to `ListCrownJewels`, added pagination parameters, and ensured uniform handling of the Crown Jewel resource across different parts of the codebase. Updated relevant comments for better code understanding.

jakule commented on api/types/crownjewel/crownjewel.go:
```diff
@@ -0,0 +1,65 @@
+	"github.com/gravitational/teleport/api/utils"
+)
+
+var _ types.Resource = &CrownJewel{}
+
+// CrownJewel is a resource that represents the crown jewel resource.
+type CrownJewel struct {
```
Sure, I can update it. But our implementation are really getting messy as I copied the pattern from access list and discovery config that AFAIU were a "popular pattern" last year.

jakule committed:
Removed CrownJewel converter, updated CrownJewel references

This commit removes the CrownJewel converter and updates references from the old CrownJewel type to the new protobuf-defined CrownJewel type, streamlining the codebase. The commit also disables CrownJewel deletion since it is not supported yet. Some minor errors were fixed regarding CrownJewel resource names.

jakule committed:
Implement Crown Jewel deletion functionality

The crown jewel deletion feature has been enabled by uncommenting and simplifying the code block. Now, the delete action is directly performed on the specific 'crown jewel' mentioned based on its name. The previously unsupported action returns an informative success message after the deletion.

jakule committed:
Refactored function 'allCombinations' and added authorization checks

Moved 'allCombinations' function into 'utils' package, renamed it to 'Combinations', and updated its references in 'dbobjectimportrulev1' and 'dbobjectv1' tests. Additionally, introduced admin authorization checks to 'CreateCrownJewels', 'UpdateCrownJewels', and 'DeleteCrownJewels'. Created corresponding tests for these changes.

jakule committed:
Add CrownJewel service to cache tests

The CrownJewel service has been added to the cache test file. The service creation is implemented and assigned to new fields in multiple methods within the cache tests. A new test function `TestCrownJewel` has also been added to validate CRUD operations for the service.

jakule committed:
Add access path changed event and update Crown Jewel tests

Added a new event code for when an access path changes in the access graph, and introduced new API calls to handle these situations. Also, updated crown jewel related tests and fixed copyright year in several files.

jakule committed:
Update protobuf import path

The protobuf import path was updated from "github.com/golang/protobuf/proto" to "google.golang.org/protobuf/proto" to ensure correct use and to prevent potential issues, while maintaining functionality in crown_jewels.go.

jakule committed:
Refactor CrownJewel and AccessGraph services

This commit optimizes the ListCrownJewelsResponse in the CrownJewel service by directly assigning the response instead of using an extra loop. Additionally, the EventsStream in the AccessGraph service and related tests have been updated to EventsStreamV2, reflecting the implementation of new versions of these methods.

jakule committed:
Add CrownJewel resource in access controls and tests

In both the cache test and access control files, CrownJewel resource has been included. This provides the system's ability to manage and test CrownJewel resources. Additionally, it allows the handling of CrownJewel resources in client events.

jakule committed:
Simplify implementation

jakule committed:
Update copyright dates and add license headers

Updated the copyright dates in several files to reflect the new year. Additionally, license headers have been added to some files where they were previously missing, ensuring the proper copyrights and GPL notice are visible.

jakule committed:
Refactor CrownJewel codebase for better readability

Fixed various typographical errors in CrownJewel's related functions and removed a TODO comment. Improved function and function calls' names that were incorrectly referring to different modules, making the code easier to read and understand across different CrownJewel services.

jakule committed:
Add validation for Teleport and AWS matchers in service

The validation has been added for Teleport and AWS matchers in the service layer to ensure that essential details are not missed when setting matchers. This includes checks to ensure that kinds and labels for Teleport, and types and tags for AWS matchers are properly set.

jakule committed:
gci

jakule commented on lib/services/local/crown_jewels_test.go:
```diff
@@ -0,0 +1,49 @@
+	require.True(t, proto.Equal(obj, newObj), "messages are not equal")
+}
+
+//TODO(jaku): add more tests
```
My bad, I added this test in a different branch by committed here 🤦

jakule commented on tool/tctl/common/resource_command.go:
```diff
@@ -911,6 +913,28 @@ func (rc *ResourceCommand) createKubeCluster(ctx context.Context, client *auth.C
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if _, err := client.CrownJewelsClient().CreateCrownJewel(ctx, crownJewel); err != nil {
+		if trace.IsAlreadyExists(err) {
+			if !rc.force {
+				return trace.AlreadyExists("crown jewel %q already exists", crownJewel.GetMetadata().GetName())
+			}
+			if _, err := client.CrownJewelsClient().UpdateCrownJewel(ctx, crownJewel); err != nil {
+				return trace.Wrap(err)
+			}
+			fmt.Printf("crown jewel %q has been updated\n", crownJewel.GetMetadata().GetName())
+			return nil
+		}
+		return trace.Wrap(err)
+	}
```
It worked when I tested it and I also call update here when create fails.

jakule committed:
Add CrownJewel update functionality and refactor related services

This commit adds the ability to update a CrownJewel resource within the ResourceCommand. It also includes refactoring of related services, condensed code logic, and improved utilization of existing methods for marshaling/unmarshaling CrownJewel. Specific changes in service.go and crown_jewels.go files have reduced code redundancy and enhanced the code structure.

jakule committed:
Relocate and update crown_jewels_test file

This commit moves the crown_jewels_test from the local to the services library and adds test cases for unmarshalling Crown Jewels data. Also, import statements for labelv1 and utils were introduced to support these changes.

jakule committed:
Update crown jewels test with AWS matchers

Revised the crown jewels test file by including AWS matchers section in the CrownJewel object. The change ensures to specifically match AWS infrastructure following certain parameters, thus enhancing the specificity and flexibility of our matchers. Apart from this, minor reordering of objects has also been done for better code organization.

jakule committed:
Added upsertCrownJewel function and related proto message

The commit adds the UpsertCrownJewel function in the service.go file, which enables upsert operations on crown jewel resources. Alongside, the UpsertCrownJewelRequest proto message is added to the crownjewel_service.pb.go file to handle such requests.

jakule committed:
Refactor CrownJewelsClient to use upsert

This change refactors the CrownJewelsClient to use the upsert method when the force flag is on. The code has been streamlined by eliminating the need for 'already exist' error checking and separate create and update methods.

jakule committed:
Add GetCrownJewel functionality and refactor validation

A GetCrownJewel function has been added to the Service, which retrieves a specific CrownJewel resource. Also, the CrownJewel validation process has been refactored, moving the validateCrownJewel function from service.go into the crownjewel library and renaming it to ValidateCrownJewel. These changes improve modularity and code organization.

jakule committed:
Add VerbCreate check and test for CrownJewel validation

The checkAccessToKind function now also checks for the VerbCreate permission. Additionally, a new test file, object_test.go, was added to validate the functionality of the CrownJewel object. This includes various validation checks such as "NilCrownJewel", "ValidCrownJewel", "MissingMatchers", among others.

jakule committed:
Refactor CrownJewel resource handling and validation

The codebase was updated to improve CrownJewel resources handling. This includes modifications to list, get, create, update, upsert, and delete functions in several files. Additionally, a validation function has been implemented for CrownJewel objects, ensuring required fields are present. Depreciated logger service has also been removed.

jakule commented on lib/auth/crownjewel/crownjewelv1/service.go:
```diff
@@ -0,0 +1,250 @@
+		return nil, trace.Wrap(err)
+	}
+
+	rsp, nextToken, err := s.backend.ListCrownJewels(ctx, req.PageSize, req.PageToken)
```
Changed, I also updated the `Get` method to read from cache

jakule committed:
Add cache implementation to CrownJewel service

The CrownJewel service has been enhanced by integrating it with cache. The main changes include adding "Cache" as part of ServiceConfig and using cache to fetch crown jewels instead of directly using the backend services. The cache interface with the methods "ListCrownJewels" and "GetCrownJewel" has been added to handle these operations.

jakule commented on lib/services/local/crown_jewels.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+func (s *CrownJewelsService) UpdateCrownJewel(ctx context.Context, crownJewel *crownjewelv1.CrownJewel) (*crownjewelv1.CrownJewel, error) {
+	r, err := s.service.UpdateResource(ctx, crownJewel)
```
I'll update it after your PR is merged.

jakule replied:
Updated.

jakule committed:
Add Cache to service configuration in crownjewelv1 tests

In this commit, we added the Cache field to the service configuration used in the tests for the crownjewelv1 service. This ensures that the service configuration used in the tests matches the expected real-world configuration.

jakule committed:
Update CrownJewelsService to use ConditionalUpdateResource

This commit changes the CrownJewelsService's `UpdateCrownJewel` method to use the `ConditionalUpdateResource` method instead of the `UpdateResource` method. This change will provide a more appropriate update strategy for the specific needs of the CrownJewelsService.

espadolini commented on lib/auth/crownjewel/crownjewelv1/service_test.go:
```diff
@@ -0,0 +1,221 @@
+	return out
+}
+
+// callMethod calls a method with given name in the DatabaseObjectService service
```
```suggestion
// callMethod calls a method with given name in the CrownJewelService service.
```

espadolini commented on lib/auth/crownjewel/crownjewelv1/service_test.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+// callMethod calls a method with given name in the DatabaseObjectService service
+func callMethod(t *testing.T, service *Service, method string) error {
```
```suggestion
func callMethod(t *testing.T, service crownjewelv1.CrownJewelServiceServer, method string) error {
```

jakule replied:
This breaks compilation and I don't want to update all types now, as we do that in all other services.

espadolini commented on lib/auth/crownjewel/crownjewelv1/service_test.go:
```diff
@@ -0,0 +1,221 @@
+
+// callMethod calls a method with given name in the DatabaseObjectService service
+func callMethod(t *testing.T, service *Service, method string) error {
+	for _, desc := range crownjewelv1.CrownJewelService_ServiceDesc.Methods {
```
The doc for `ServiceName_ServiceDesc` states

```
// It's only intended for direct use with grpc.RegisterService,
// and not to be introspected or modified (even as a copy)
```

but I can't find a useful way to make use of `protoreflect`/`protoregistry`, and I'm not entirely sure if going through `reflect` is better than going through an exported var that Must Not Be Used (and that other projects are indubitably also using).

<details>
<summary>for posterity</summary>

```go
	m := reflect.ValueOf(service).MethodByName(method)
	if !m.IsValid() {
		require.FailNow(t, "method %v not found", method)
	}
	r := m.Call([]reflect.Value{
		reflect.ValueOf(context.Background()),
		reflect.New(m.Type().In(1).Elem()),
	})
	err := r[1].Interface()
	if err == nil {
		return nil
	}
	return err.(error)
```

</details>

jakule replied:
>and not to be introspected or modified

Reflect sounds like introspecting 🤔 I'll leave the current version as it looks "safer" IMO

espadolini commented on lib/auth/crownjewel/crownjewelv1/service_test.go:
```diff
@@ -0,0 +1,221 @@
+
+func (f fakeChecker) CheckAccessToRule(_ services.RuleContext, _ string, resource string, verb string) error {
+	if resource == types.KindCrownJewel {
+		for _, allowedVerb := range f.allowedVerbs {
+			if allowedVerb == verb {
+				return nil
+			}
+		}
```
```suggestion
		if slices.Contains(f.allowedVerbs, verb) {
			return nil
		}
```

espadolini commented on lib/auth/crownjewel/object.go:
```diff
@@ -0,0 +1,105 @@
+	return cj, nil
+}
+
+// ValidateCrownJewel validates the CrownJewel object.
```
I think it's useful if we start documenting this sort of details.

```suggestion
// ValidateCrownJewel validates the CrownJewel object without modifying.
```

espadolini commented on lib/auth/crownjewel/object.go:
```diff
@@ -0,0 +1,105 @@
+//   - Matcher.Name or Matcher.Labels
+func ValidateCrownJewel(jewel *crownjewelv1.CrownJewel) error {
+	switch {
+	case jewel == nil:
```
for this and the later parts: a `nil` pointer protobuf message is a valid read-only message that has all fields zeroed; should we start using getters whenever possible? we've kind of avoided that for gogoproto, which doesn't always generate getters and can store submessages as structs, but that's not how the convention goes for protobuf

espadolini commented on lib/auth/crownjewel/object.go:
```diff
@@ -0,0 +1,105 @@
+				return trace.BadParameter("teleport matcher name or labels must be set")
+			}
+
+			if len(matcher.GetLabels()) > 0 {
+				for _, label := range matcher.GetLabels() {
```
Unnecessary check

espadolini commented on lib/auth/crownjewel/object.go:
```diff
@@ -0,0 +1,105 @@
+				return trace.BadParameter("aws matcher arn or tags must be set")
+			}
+
+			if len(matcher.GetTypes()) > 0 {
+				for _, label := range matcher.GetTags() {
```
Typo, but even if the check was for GetTags it'd be unnecessary 🤣

espadolini commented on lib/cache/cache.go:
```diff
@@ -2811,7 +2823,33 @@ func (c *Cache) GetDiscoveryConfig(ctx context.Context, name string) (*discovery
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+	return rg.reader.GetCrownJewel(ctx, name)
```
Would it make sense to read through the cache on a `NotFound`? We do that for some resources like roles, for example.

espadolini commented on lib/services/crown_jewel.go:
```diff
@@ -0,0 +1,87 @@
+		object = proto.Clone(object).(*crownjewelv1.CrownJewel)
+		object.Metadata.Revision = ""
+	}
+	data, err := utils.FastMarshal(object)
```
no protojson? 🥺

jakule committed:
Replace 'Cache' with 'Reader' in crownjewel service & remove DeleteAllCrownJewelsRequest

The 'Cache' in crownjewel service has been replaced with 'Reader' to improve code readability and to match with its functionality. Also, the DeleteAllCrownJewelsRequest functionality was removed as it posed a high risk potential for data loss and it wasn

r0mant reviewed (approved):
Bot.

jakule committed:
Update tests for Crown Jewel service

This commit updates the test suite of the Crown Jewel service. An object is fetched from the backend before updating its expiry time, ensuring the revision is populated. In addition, a new test is introduced to handle the scenario where the revision is missing when updating a Crown Jewel.

jakule committed:
Refactor protobuf marshaling with utils

This commit replaces the usage of the `protojson` method for marshaling and unmarshaling protobuf objects in `crown_jewel.go` with the `FastMarshal` and `FastUnmarshal` methods from `utils`.
protojson is failing to unmarshal resources.

jakule committed:
Update CrownJewel test and serialization methods

Updated the test condition in crown_jewels_test.go to use Proto.Equal for object comparisons. Also, swapped out the FastMarshal and FastUnmarshal methods with protojson's Marshal and Unmarshal in the CrownJewels service.

jakule committed:
Add missing license
